src/analysis/processing/qgsalgorithmorderbyexpression.cpp

```diff
@@ -0,0 +1,106 @@
+/***************************************************************************
+                         qgsalgorithmorderbyexpression.h
+                         ---------------------
+    begin                : November 2017
+    copyright            : (C) 2017 by Etienne Trimaille
+    email                : etienne dot trimaille at gmail dot com
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+
+#include "qgsalgorithmorderbyexpression.h"
+#include "qgsfeaturerequest.h"
+
+
+///@cond PRIVATE
+
+QString QgsOrderByExpressionAlgorithm::name() const
+{
+  return QStringLiteral( "orderbyexpression" );
+}
+
+QString QgsOrderByExpressionAlgorithm::displayName() const
+{
+  return QObject::tr( "Order by expression" );
+}
+
+QStringList QgsOrderByExpressionAlgorithm::tags() const
+{
+  return QObject::tr( "orderby,sort,expression,field" ).split( ',' );
+}
+
+QString QgsOrderByExpressionAlgorithm::group() const
+{
+  return QObject::tr( "Vector general" );
+}
+
+void QgsOrderByExpressionAlgorithm::initAlgorithm( const QVariantMap & )
+{
+  addParameter( new QgsProcessingParameterFeatureSource( QStringLiteral( "INPUT" ), QObject::tr( "Input layer" ) ) );
+  addParameter( new QgsProcessingParameterExpression( QStringLiteral( "EXPRESSION" ), QObject::tr( "Expression" ), QVariant(), QStringLiteral( "INPUT" ) ) );
+  addParameter( new QgsProcessingParameterBoolean( QStringLiteral( "ASCENDING" ), QObject::tr( "Ascending" ), true ) );
+  addParameter( new QgsProcessingParameterBoolean( QStringLiteral( "NULLS_FIRST" ), QObject::tr( "Nulls first" ), false ) );
+
+  addParameter( new QgsProcessingParameterFeatureSink( QStringLiteral( "OUTPUT" ), QObject::tr( "Output layer" ) ) );
+}
+
+QString QgsOrderByExpressionAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "This algorithm sorts a vector layer according to an expression." );
```

Comment:
I think here should also a note be added, that it's up to each provider if this order is preserved and that relying on this is often exploiting undocumented behavior. (in fact, I think this algorithm will only unleash its full potential within the specs once we have feature pipelines in place).

Reply:
Thanks @m-kuhn I saw your comment about ordering yesterday https://gis.stackexchange.com/questions/263130/render-polygons-ordered-by-size/263133#263133

I tried on some files (geojson) and it seems to work, but you are saying it's not expected for now? Even with postgres provider? Do you have an idea of which providers might respect the order which providers don't?

Reply:
I think many will do it to some degree, especially simple transfer files like geojson, but I think it's exploiting undocumented behavior and may change depending on context, indexing, ...